### ISSUE TITLE
timeslist does not expire

### ISSUE THREAD

**Issue by mperham** [davydovanton/sidekiq-statistic#73]
https://github.com/davydovanton/sidekiq-statistic/blob/master/lib/sidekiq/statistic/middleware.rb#L40

The timeslist should be expired after some amount of time and (I'd suggest) the times aggregated into a much more compact form.  The Web UI could store hourly aggregates per worker.

If I were implementing this, I'd compact the timeslists hourly.  The compact form could store:

```
execution count
average time
standard deviation
98th percentile
etc
```

So instead of a list of thousands of numbers, it would be converted into a hash with 4-6 numbers per hour.  You could expire the hashes after one month, giving the user a nice recent history to compare.

**Comment by antulik** [comment]
+1 constantly growing redis memory usage became an issue for us due to sidekiq-statistic

**Comment by felixbuenemann** [comment]
Maybe hyperloglog can be used to aggregate execution counts over longer periods. Redis' hyperloglog is exact for smaller counts and can be merged to aggregate data. So it would be possible to store counts per day/hour/whatever and then aggregate to bigger intervals. Eg. daily merge per hour to per day, per day to week and merge to total.

This only works for counting things and not for storing durations and all the possible keys have to be stores separately. But it's probably OK to have average execution time over a bounded period (eg. last month) and execution count per status unbounded using hll.

**Comment by eggie5** [comment]
+1 I just tracked down our out of control memory growth to this!

**Comment by shaicoleman** [comment]
+1, Redis killed by the Linux Out of Memory Killer.

As a workaround, I clear all the timeslist on a regular basis like this:
`Sidekiq.redis { |connection| connection.keys('*:timeslist').each { |key| connection.del(key) } }`

**Comment by mperham** [comment]
@davydovanton I've put this gem on the "troublesome" list and removed it from the Related Projects wiki page since this bug hasn't been fixed.

**Comment by eggie5** [comment]
@mperham yeah I think that would save everybody a lot of trouble.

**Comment by jacek213** [comment]
That's a major issue, anyone tried to fixed that? I have to stop using this lib until it's fixed.

**Comment by garethson** [comment]
Lots more could be done to re-vamp how times are calculated and stored, but I think the solution I've got up at https://github.com/davydovanton/sidekiq-statistic/pull/107 could bring this gem back to stable and performant for those running a lot of jobs.

**Comment by MrJoy** [comment]
@garethson Does that mean this issue should be be closed, with a follow-up issue to improve storage efficiency?  Or is there still a problem of unbounded memory growth after your patch?

**Comment by garethson** [comment]
@MrJoy This did seem to address the issue we were having, yes. It safely drops (very) old history once an upper bound is reached, so perhaps not the most elegant solution, but works for us while still rendering the tool very useful.

**Comment by wenderjean** [comment]
I'm closing this issue considering we already have a feature to get rid of pretty old info on Redis.

**Comment by drn** [comment]
Hi @wenderjean — can you clarify which feature you're referring to?

**Comment by MrJoy** [comment]
He's referring to expiration times.  Redis can act as an LRU cache either globally, or explicitly per-value when an expiration is set.

**Comment by drn** [comment]
Thanks for clarifying, @MrJoy

This may have changed recently, but my understanding is that there is no option to specify a global TTL in Redis and it must be specified per key. It looks like Redis 4.0 introduced some new functionality — https://redis.io/topics/lru-cache, but the eviction policies seem geared for using Redis as a cache rather than a queue.

Rather than having the users of this library adjust global Redis configuration or expiring keys generated by this library explicitly, it seems like it would make sense for the library to handle setting TTLs for these keys directly.

Feel free to leave this closed (my organization no longer uses the gem), but I do think that this library could benefit from handling expirations more explicitly either by default or through configuration.

Cheers 👋

**Comment by wenderjean** [comment]
Sorry for the delay @drn.

I was actually referring to this piece of code: https://github.com/davydovanton/sidekiq-statistic/blob/master/lib/sidekiq/statistic/statistic/metrics/store.rb#L52

It was added in the past and since then we haven't seen activities here, that's the reason I closed the issue.

We should probably have other approaches to handle that that I really think we can discuss further so, we may want to create a new issue to change what we actually have if that's the case.
